### ISSUE TITLE
minimap2 short-read (-ax sr) much slower than bwa mem on some WGS

### ISSUE THREAD

**Issue by wilsonte-umich** [lh3/minimap2#1180]
Minimap2 is said to be 3 times as fast as BWA-MEM for Illumina short reads.
I consistently find minimap2 much slower than BWA-MEM on some WGS paired reads.

I compared my current sample to SRA run DRR518045.  My reads are derived
from an input where mouse and yeast DNA were present together, 
DRR518045 is standard mouse WGS.

Ordinarily, I align our mouse-yeast (MY) data to a composite
genome with concatenated mm10 and sacCer3 (mm10_sacCer3). For comparison,
I aligned 500K read pairs from either sample, to either mm10 or mm10_sacCer3, with each aligner.

```
==========================================================================
BWA_VERSION       0.7.17
MINIMAP2_VERSION  2.27
==========================================================================
bwa mem -Y -t 8 genome.fa fastq_1.gz fastq_2.gz
minimap2 -ax sr -Y -t 8 genome.fa fastq_1.gz fastq_2.gz 
==========================================================================
minimap2 indexing of mm10 (includes unplaced sequences, etc.)
--------------------------------------------------------------------------
[M::mm_idx_gen::50.238*1.64] collected minimizers
[M::mm_idx_gen::58.455*2.25] sorted minimizers
[M::main::58.455*2.25] loaded/built the index for 66 target sequence(s)
[M::mm_mapopt_update::58.455*2.25] mid_occ = 1000
[M::mm_idx_stat] kmer size: 21; skip: 11; is_hpc: 0; #seq: 66
[M::mm_idx_stat::63.280*2.16] distinct minimizers: 339456585 (96.18% are singletons); average occurrences: 1.316; average spacing: 6.113; total length: 2730871774
--------------------------------------------------------------------------
minimap2 indexing of mm10_sacCer3 (canoniocal chromosomes only)
--------------------------------------------------------------------------
[M::mm_idx_gen::48.458*1.61] collected minimizers
[M::mm_idx_gen::54.652*2.14] sorted minimizers
[M::main::54.652*2.14] loaded/built the index for 39 target sequence(s)
[M::mm_mapopt_update::54.652*2.14] mid_occ = 1000
[M::mm_idx_stat] kmer size: 21; skip: 11; is_hpc: 0; #seq: 39
[M::mm_idx_stat::61.442*2.01] distinct minimizers: 341254243 (96.20% are singletons); average occurrences: 1.313; average spacing: 6.112; total length: 2737694774
```

| Data      |  Genome      | Aligner  | Walltime |  Reads/Sec |  primary mapped |  supplementary |  properly paired |
|-----------|--------------|----------|---------:|-----------:|-----------:|-----------:|-----------:|
| DRR518045 | mm10         | bwa      |    1m56s |  10823     | 99.73% | 9431 | 97.21% |
| DRR518045 | mm10         | minimap2 |    1m51s |   9127     | 98.90% | 4767 | 95.78% |
| DRR518045 | mm10_sacCer3 | bwa      |    1m38s |  12677     | 99.72% | 9313 | 97.29% |
| DRR518045 | mm10_sacCer3 | minimap2 |    1m46s |   9476     | 98.85% | 4618 | 95.74% |
| MY | mm10         | bwa      |  2m37s  |  6769 | 77.27% |  90550 | 70.43% |
| MY | mm10         | minimap2 | 30m17s  |   550 | 72.97% | 114882 | 64.34% |
| MY | mm10_sacCer3 | bwa      |  2m8s   |  9214 | 98.89% |  96993 | 91.39% |
| MY | mm10_sacCer3 | minimap2 | 27m4s   |   616 | 98.09% | 113922 | 88.08% |

I never see `minimap2` is 3 times faster than `bwa mem`.

Most importantly, on the MY dataset, regardless of genome,
`minimap2` is ~10 times slower than `bwa mem`. MY is 
marginally slower with `bwa mem` as compared to DRR518045.

Thus, very slow alignment is seen specifically with the combination
of MY reads and minimap2.

There is apparently something unhappy about MY, as even with the composite genome
the % properly paired is notably lower and the supplementary alignments quite high.

I will continue to explore read properties, but regardless, would 
appreciate insights as to why `minimap2` becomes uniquely so much slower based
on input reads, and what parameters would be most useful for speeding it back up.

**Comment by lh3** [comment]
See http://lh3.github.io/2018/04/02/minimap2-and-the-future-of-bwa

**Comment by wilsonte-umich** [comment]
I want to added a comment for anyone who lands here that I found a key factor limiting minimap2 alignment speed on my short reads.  

As I compared different Illumina NovaSeqX runs - some with much lower alignment speed - I noticed a lot of poly-G and poly-C reads (a topic for another forum why so many were in some Illumina libraries). I surmised minimap2 had a particularly hard timing mapping them.

I implemented a poly-X tail filter (using fastp) - it trims not only poly-G tails resulting from Illumina no-color bases, but also poly-C reads contaminating some of my inputs. There is also significant base quality filtering and adapter trimming, but the key factor was the poly-X filter.

`minimap2` ran MUCH faster after trimming poly-X. One library with higher poly-X contamination **aligned at least 20-fold faster than without poly-X filtering**! 

In a pipe from `fastp` to `minimap2`, `top` process monitoring showed minimap2 was strongly rate limiting without poly-X filtering. With poly-X filtering, fastp used a significant portion of cpu, indicating minimap2 now aligned reads consistently fast without bogging down on offending reads.

I find that with proper read preparation minimap2 is a great and fast short read aligner but slows dramatically the more garbage reads you give it.
